crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -325,6 +324,57 @@ impl CumulocityConverter {
         })
     }
 
+    pub async fn try_register_source_entities(
+        &mut self,
+        message: &MqttMessage,
+    ) -> Result<Vec<PendingEntityData>, ConversionError> {
+        if let Ok((source, channel)) = self.mqtt_schema.entity_channel_of(&message.topic) {
+            match channel {
+                Channel::EntityMetadata => {
+                    if let Ok(register_message) = EntityRegistrationMessage::try_from(message) {
+                        return self
+                            .try_register_entity_with_pending_children(register_message)
+                            .await;
+                    }
+                    Err(anyhow!(
+                        "Invalid entity registration message received on topic: {}",
+                        message.topic.name
+                    )
+                    .into())
+                }
+                _ => {
+                    // if the target entity is unregistered, try to register it first using auto-registration
+                    if self.entity_store.get(&source).is_none()
+                        && self.config.enable_auto_register
+                        && source.matches_default_topic_scheme()
+                    {
+                        let auto_registered_entities = self.try_auto_register_entity(&source)?;
+                        Ok(auto_registered_entities
+                            .into_iter()
+                            .map(|reg_msg| reg_msg.into())
+                            .collect())
+                    } else {
+                        Ok(vec![])
```

Comment:
As the message is related to some unknown entity, I would expect an error here.

Reply:
I didn't get that. This `else` case is for the case where the target topic id is **present** in the entity store. So, there is nothing left to be registered and hence returning empty vector.

Reply:
> This `else` case is for the case where the target topic id is **present** in the entity store.

Not only. This `else` case also covers the case where the target topic id  is unknown *but* auto-registration is off or the topic id is not along the default scheme (`device/<name>//`).

Reply:
> > This `else` case is for the case where the target topic id is **present** in the entity store.
> 
> Not only. This `else` case also covers the case where the target topic id is unknown _but_ auto-registration is off or the topic id is not along the default scheme (`device/<name>//`).

Correct. In those cases as well, we're supposed to just cache that message in the pending entity store to be processed later. That caching logic was in the `try_convert_data_message` earlier. I've now moved it here, with detailed commentary, to avoid this confusion. Hope the logic is clearer now.

Reply:
Clearer